File: Solid Principles/O - Open to extension and close to modification/CorrectExample.java
import java.util.*;
import java.io.*;

class Marker{
    private String name;
    private String color;
    private Integer price;
    private Integer year;

    public int getPrice(){
        return price;
    }
}

class Invoice{
    Integer quantity;
    Marker marker;

    Invoice(Integer quantity, Marker marker){
        this.quantity = quantity;
        this.marker = marker;
    }

    int calculateTotal(){
        return marker.getPrice() * quantity;
    }
}

// Create an interface to InvoiceDao

interface InvoiceDao{
    void save(Invoice invoice);
}

class DatabaseInvoiceDao implements InvoiceDao{

    @Override
    public void save(Invoice invoice) {
        // Add DB Related Operations
    }
    
}

class FileInvoiceDao implements InvoiceDao{

    @Override
    public void save(Invoice invoice) {
        // Add File Saving operations.
    }
    
}

public class CorrectExample{
    public static void main(String[] args) {
        
    }
}

File: Solid Principles/O - Open to extension and close to modification/correct_example.py
class Marker:
    name : str
    color : str
    year : int 
    price : int

    def __init__(self, name, color, year, price):
        self.name = name
        self.color = color
        self.year = year
        self.price = price

class Invoice:
    __marker : Marker
    __quantity : int

    def __init__(self, marker, quantity):
        self.__marker = marker
        self.__quantity = quantity

    def calculate_total(self):
        return self.__marker.price * self.__quantity

class InvoiceDao:
    invoice : Invoice

    def __init__(self, invoice):
        self.invoice = invoice
        pass

    def save(self):
        pass

class DatabaseInvoiceDao(InvoiceDao):

    def __init__(self, invoice):
        super().__init__(invoice) 

    def save(self):
        # Perform Database Operations
        pass
    
class FileInvoiceDao(InvoiceDao): 

    def __init__(self, invoice):
        super().__init__(invoice) 

    def save(self):
        # Perform File Operations
        pass
